In [1]:
import pandas as pd

In [2]:
cd D:\HKUST\Foundations of Data Analysis\group project\tap4fun竞赛数据

D:\HKUST\Foundations of Data Analysis\group project\tap4fun竞赛数据


In [3]:
data = pd.read_csv("tap_fun_train.csv", parse_dates=True)
data_test = pd.read_csv("tap_fun_test.csv", parse_dates=True)

check duplicated user_id

In [4]:
data_id = pd.DataFrame(data['user_id'],columns=['user_id'])
data_test_id = pd.DataFrame(data_test['user_id'],columns=['user_id'])
pd.merge(data_id, data_test_id, on = 'user_id')

Empty DataFrame
Columns: [user_id]
Index: []

In [5]:
data.columns

Index(['user_id', 'register_time', 'wood_add_value', 'wood_reduce_value',
       'stone_add_value', 'stone_reduce_value', 'ivory_add_value',
       'ivory_reduce_value', 'meat_add_value', 'meat_reduce_value',
       ...
       'pvp_battle_count', 'pvp_lanch_count', 'pvp_win_count',
       'pve_battle_count', 'pve_lanch_count', 'pve_win_count',
       'avg_online_minutes', 'pay_price', 'pay_count', 'prediction_pay_price'],
      dtype='object', length=109)

time distribution of user registration

In [6]:
data['register_time_month'] = data.register_time.str[:7]
data['register_time_day'] = data.register_time.str[6:10]

In [7]:
data_month_df = pd.DataFrame(data['register_time_month'].value_counts()).sort_index()
print(data_month_df)
data_day_df = pd.DataFrame(data['register_time_day'].value_counts()).sort_index()
print(data_day_df)

         register_time_month
2018-01               390420
2018-02              1632463
2018-03               265124
      register_time_day
1-26              70250
1-27              70417
1-28              79227
1-29              63803
1-30              50201
1-31              56522
2-01              83245
2-02              60173
2-03              51659
2-04              60421
2-05              60998
2-06              57203
2-07              71576
2-08              72402
2-09              50143
2-10              53521
2-11              54014
2-12              52231
2-13              50638
2-14              54419
2-15              78707
2-16              56355
2-17              44477
2-18              59447
2-19             117311
2-20              92860
2-21              43720
2-22              42110
2-23              44635
2-24              45648
2-25              49835
2-26              42647
2-27              39140
2-28              42928
3-01              36226
3-02              42

In [8]:
from pyecharts import Line, Grid

ERROR:lml.utils:failed to import pyecharts_snapshot
Traceback (most recent call last):
  File "D:\software\anoconda\anaconda\lib\site-packages\lml\utils.py", line 43, in do_import
    plugin_module = __import__(plugin_module_name)
ModuleNotFoundError: No module named 'pyecharts_snapshot'


In [9]:
line1 = Line("number of users who registered-monthly")
line1.add("No. users", data_month_df.index, data_month_df['register_time_month'], mark_line=["average"], mark_point=["max", "min"])

line2 = Line("number of users who registered-daily",title_top="50%")
line2.add("No. users", data_day_df.index, data_day_df['register_time_day'], mark_line=["average"], mark_point=["max", "min"])

grid = Grid(width = 1000, height = 1000)
grid.add(line1, grid_bottom="60%")
grid.add(line2, grid_top="60%")
grid.render()
grid

number of users who paid within 7 days

In [10]:
import copy

In [11]:
data_pay_7 = copy.copy(data[data['pay_price']>0])
print(data_pay_7.shape)   # (41439, 111)
print(data_pay_7.shape[0]/data.shape[0])  # 0.018111395638212645

(41439, 111)
0.018111395638212645


the change of the ratio of users who paid with time

In [12]:
data_pay_7_day_df = pd.DataFrame(data_pay_7['register_time_day'].value_counts()).sort_index()
# print(data_pay_7_day_df)
data_pay_7_day_df.rename(columns={'register_time_day':'pay_register_time_day'}, inplace = True)
data_day_count = pd.concat([data_pay_7_day_df, data_day_df], axis=1)
# print(data_day_count)
data_day_count['pay_percent'] = data_day_count['pay_register_time_day']/data_day_count['register_time_day']
print(data_day_count)

      pay_register_time_day  register_time_day  pay_percent
1-26                   1636              70250     0.023288
1-27                   1519              70417     0.021571
1-28                   1513              79227     0.019097
1-29                   1498              63803     0.023479
1-30                   1236              50201     0.024621
1-31                   1032              56522     0.018258
2-01                   1126              83245     0.013526
2-02                   1393              60173     0.023150
2-03                   1201              51659     0.023249
2-04                   1155              60421     0.019116
2-05                   1356              60998     0.022230
2-06                    992              57203     0.017342
2-07                    976              71576     0.013636
2-08                   1016              72402     0.014033
2-09                   1111              50143     0.022157
2-10                   1039             

In [13]:
from pyecharts import Overlap

In [15]:
data_day_count

pay_register_time_day  register_time_day  pay_percent
1-26                   1636              70250     0.023288
1-27                   1519              70417     0.021571
1-28                   1513              79227     0.019097
1-29                   1498              63803     0.023479
1-30                   1236              50201     0.024621
1-31                   1032              56522     0.018258
2-01                   1126              83245     0.013526
2-02                   1393              60173     0.023150
2-03                   1201              51659     0.023249
2-04                   1155              60421     0.019116
2-05                   1356              60998     0.022230
2-06                    992              57203     0.017342
2-07                    976              71576     0.013636
2-08                   1016              72402     0.014033
2-09                   1111              50143     0.022157
2-10                   1039              53521     0.019413
2-11                    962              54014     0.017810
2-12                   1206              52231     0.023090
2-13                    841              50638     0.016608
2-14                    703              54419     0.012918
2-15                    687              78707     0.008729
2-16                    989              56355     0.017549
2-17                    749              44477     0.016840
2-18                    784              59447     0.013188
2-19                   1052             117311     0.008968
2-20                    839              92860     0.009035
2-21                   1021              43720     0.023353
2-22                    834              42110     0.019805
2-23                   1007              44635     0.022561
2-24                    866              45648     0.018971
2-25                    855              49835     0.017157
2-26                   1034              42647     0.024246
2-27                    695              39140     0.017757
2-28                    740              42928     0.017238
3-01                    666              36226     0.018385
3-02                   1032              42775     0.024126
3-03                    904              48970     0.018460
3-04                    994              50989     0.019494
3-05                   1169              44726     0.026137
3-06                   1011              41438     0.024398

In [14]:
line3 = Line()
line3.add("number of users who registered", data_day_count.index, data_day_count['register_time_day'], mark_line=["average"], mark_point=["max", "min"])

line4 = Line()
line4.add("number of users who paid within 7 days", data_day_count.index, data_day_count['pay_register_time_day'], mark_line=["average"], 
          mark_point=["max", "min"], yaxis_max=3000)

overlap = Overlap()
# default: no new x, y axis, the index of x,y axis are 0
overlap.add(line3)
overlap.add(line4, yaxis_index=1, is_add_yaxis=True)
overlap.render()

overlap

In [15]:
from pyecharts import Bar, Overlap

In [16]:
line3 = Line()
line3.add("number of users who registered", data_day_count.index, data_day_count['register_time_day'], mark_line=["average"], mark_point=["max", "min"])

bar = Bar()
bar.add("raio of users who paid within 7 days", data_day_count.index, data_day_count['pay_percent'], yaxis_max=0.1)

overlap = Overlap()
overlap.add(line3)
overlap.add(bar, yaxis_index=1, is_add_yaxis=True)
overlap.render()

overlap

How much did they pay

In [17]:
data_pay_45 = copy.copy(data[data['prediction_pay_price']!=0])
print(data_pay_45['prediction_pay_price'].describe())
print('Total amount within first 45 days:',data_pay_45['prediction_pay_price'].sum())

count    45988.000000
mean        89.213058
std        617.702040
min          0.990000
25%          0.990000
50%          3.970000
75%         20.980000
max      32977.810000
Name: prediction_pay_price, dtype: float64
Total amount within first 45 days: 4102730.1099999994


46 thousand users paid within first 45 days, and the total amount is 4.1 million. The richest (土豪) paid 33 thousand! You never understand ...

In [18]:
data_pay_7 = copy.copy(data[data['pay_price']!=0])
print(data_pay_7['pay_price'].describe())
print('Total amount within first 7 days：',data_pay_7['pay_price'].sum())  
print('Total amount within first 45 days：',data_pay_7['prediction_pay_price'].sum())

count    41439.000000
mean        29.521143
std        165.655561
min          0.990000
25%          0.990000
50%          1.990000
75%         11.970000
max       7457.950000
Name: pay_price, dtype: float64
Total amount within first 7 days： 1223326.66
Total amount within first 45 days： 3916936.11


Among 46 thousand users, 41 thousand (90.1% of 46 thousand) paid within 7 days. These 41 thousand users paid 1.2 million for the first 7 days, and paid 3.9 million for the first 45 days (95.5% of the total amoung 4.1 million)

In [19]:
data_nopay_7_pay_45 = copy.copy(data_pay_45[data_pay_45['pay_price']==0])
print(data_nopay_7_pay_45['prediction_pay_price'].describe())
print('number of users who did not pay for the first 7 days but paid later：',data_nopay_7_pay_45['prediction_pay_price'].sum())

count    4549.000000
mean       40.842823
std       238.764700
min         0.990000
25%         0.990000
50%         5.980000
75%        19.950000
max      6752.970000
Name: prediction_pay_price, dtype: float64
number of users who did not pay for the first 7 days but paid later： 185794.0


4.5 thousand users (9.9%) did not pay for the first 7 days, but pay 186 thousand within 45 days (4.5%)

In [20]:
data_pay_7_nopay_45 = copy.copy(data_pay_7[data_pay_7['pay_price']==data_pay_7['prediction_pay_price']])
print(data_pay_7_nopay_45['pay_price'].describe())
print(data_pay_7_nopay_45['pay_count'].describe())
print('Total amount within first 7 days：',data_pay_7_nopay_45['pay_price'].sum())  
print('number of users who only paid for the first 7 days：',data_pay_7_nopay_45.shape[0])

count    30130.000000
mean        11.407407
std         71.134875
min          0.990000
25%          0.990000
50%          1.980000
75%          5.980000
max       4086.520000
Name: pay_price, dtype: float64
count    30130.000000
mean         2.239728
std          2.340433
min          1.000000
25%          1.000000
50%          1.000000
75%          3.000000
max         54.000000
Name: pay_count, dtype: float64
Total amount within first 7 days： 343705.17000000004
number of users who only paid for the first 7 days： 30130


30 thousand users (65.5% of 46 thousand) only paid for the first 7 days. Total amount: 344 thousand (28.1% of the first 7 days, 8.4% of 45 days)

In [21]:
data_pay_7_pay_45 = copy.copy(data_pay_7[data_pay_7['pay_price']<data_pay_7['prediction_pay_price']])
print(data_pay_7_pay_45['pay_price'].describe())
print(data_pay_7_pay_45['pay_count'].describe())
print('Total amount within first 7 days：',data_pay_7_pay_45['pay_price'].sum())  
print('Total amount within first 45 days：',data_pay_7_pay_45['prediction_pay_price'].sum())  
print('number of users who paid within both 7 days and 45 days：',data_pay_7_pay_45.shape[0])

count    11309.000000
mean        77.780661
std        289.610458
min          0.990000
25%          2.980000
50%         15.970000
75%         45.940000
max       7457.950000
Name: pay_price, dtype: float64
count    11309.000000
mean         5.707932
std          6.470337
min          1.000000
25%          2.000000
50%          4.000000
75%          7.000000
max        105.000000
Name: pay_count, dtype: float64
Total amount within first 7 days： 879621.4900000001
Total amount within first 45 days： 3573230.9399999995
number of users who paid within both 7 days and 45 days： 11309


11 thousand users (25% of 46 thousand) paid within first 7 days and continued to pay within 45 days. Total amount 3.57 million (87.1%). The average times of payment is 5.7, meaning that many paid more than once within 7 days.

How long did they play

In [22]:
data['avg_online_minutes'].describe()

count    2.288007e+06
mean     1.020749e+01
std      3.895946e+01
min      0.000000e+00
25%      5.000000e-01
50%      1.833333e+00
75%      4.833333e+00
max      2.049667e+03
Name: avg_online_minutes, dtype: float64

75% played for less than 45 minutes.

In [23]:
data_pay = copy.copy(data[data['pay_price']!=0])
data_pay['avg_online_minutes'].describe()

count    41439.000000
mean       140.187746
std        149.972913
min          0.000000
25%         33.000000
50%         88.833333
75%        194.666667
max       1674.666667
Name: avg_online_minutes, dtype: float64

Users who paid played for 140 minutes on average. That is 20 minutes per day.

the amount of each payment

In [24]:
data_once = copy.copy(data[data['pay_count']==1])
print(data_once.shape)
data_once.groupby("pay_price")["pay_count"].sum()

(17622, 111)


pay_price
0.99     15791
1.99       610
4.99       725
9.99       407
19.99       66
49.99        7
99.99       16
Name: pay_count, dtype: int64

The minimum amount of each payment is 0.99. Other types of payment all ended with 0.99.

Evaluation criterion: the richest users may heavily bias the result. Whether succeed or not depends on the richest.

test data

In [25]:
data_test['register_time_day'] = data_test.register_time.str[6:10]

In [26]:
data_test_day_df = pd.DataFrame(data_test['register_time_day'].value_counts()).sort_index()

In [27]:
data_test_pay_7 = copy.copy(data_test[data_test['pay_price']>0])
data_test_pay_7_day_df = pd.DataFrame(data_test_pay_7['register_time_day'].value_counts()).sort_index()
data_test_pay_7_day_df.rename(columns={'register_time_day':'pay_register_time_day'}, inplace = True)
data_day_count = pd.concat([data_test_pay_7_day_df, data_test_day_df], axis=1)

In [28]:
data_day_count['pay_percent'] = data_day_count['pay_register_time_day']/data_day_count['register_time_day']

In [29]:
line3 = Line()
line3.add("number of users who registered", data_day_count.index, data_day_count['register_time_day'], mark_line=["average"], mark_point=["max", "min"])

bar = Bar()
bar.add("raio of users who paid within 7 days", data_day_count.index, data_day_count['pay_percent'], yaxis_max=0.1)


overlap = Overlap()
overlap.add(line3)
overlap.add(bar, yaxis_index=1, is_add_yaxis=True)
overlap.render()

overlap

In [30]:
data_test_pay_7 = copy.copy(data_test[data_test['pay_price']!=0])
print(data_test_pay_7['pay_price'].describe())
print('Total amount within first 7 days：',data_test_pay_7['pay_price'].sum())  

count    19549.000000
mean        28.487393
std        183.378785
min          0.990000
25%          0.990000
50%          1.990000
75%         11.970000
max      14495.220000
Name: pay_price, dtype: float64
Total amount within first 7 days： 556900.04


Similar distribution with train set

In [31]:
import sys
sys.modules[__name__].__dict__.clear()

Classification

In [32]:
import pandas as pd
import numpy as np
import copy
import matplotlib.pyplot as plt
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection  import train_test_split
from sklearn.metrics import f1_score
import time

In [45]:
data = pd.read_csv("tap_fun_train.csv", parse_dates=True)

In [47]:
data

KeyError: '_oh'

In [36]:
object_columns_df = data.select_dtypes(include=["object"])
print(object_columns_df.iloc[1])

register_time    2018-01-26 00:01:05
Name: 1, dtype: object


In [37]:
# remove register_time
data['register_time_month'] = data.register_time.str[5:7]
data['register_time_day'] = data.register_time.str[8:10]
data = data.drop(['register_time'],axis=1) 

In [38]:
data[['register_time_month','register_time_day']] = data[['register_time_month','register_time_day']].apply(pd.to_numeric)
# data=pd.DataFrame(data,dtype=np.float) 

In [39]:
# combine month and day to sort register time
data['register_time_count'] = data['register_time_month'] * 31 + data['register_time_day'] 

In [43]:
type(data)

KeyError: '_oh'

In [ ]:
# save users who paid within first 7 days
data_7_pay = copy.copy(data[data['pay_price']>0])
data_7_pay.shape

In [ ]:
data_7_pay.to_csv ("tap_fun_train_7_pay.csv")

In [ ]:
data = pd.read_csv("tap_fun_train_7_pay.csv", index_col=0, parse_dates=True)
print(data.shape) 

Assign labels
* 1: only paid within 7 days
* 0: continued to pay

In [ ]:
data['7_45_same_pay_label'] = (data['pay_price'] == data['prediction_pay_price'])
data['7_45_same_pay_label']=data['7_45_same_pay_label'].map({True:1,False:0})

data['7_45_same_pay_label'].value_counts()

In [ ]:
data = data.drop(['prediction_pay_price', 'user_id'],axis=1)
data.shape

Train the classification model

In [ ]:
label = '7_45_same_pay_label'

In [ ]:
X = data.loc[:, data.columns != label]
y = data.loc[:, data.columns == label]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.22, random_state = 0)

In [ ]:
print("Number transactions train dataset: ", len(X_train))
print("Number transactions test dataset: ", len(X_test))
print("Total number of transactions: ", len(X_train)+len(X_test))

In [ ]:
data_train = pd.concat( [X_train, y_train], axis=1 )
print(data.shape)
print('---------------------------------------')
print(data_train.shape)

In [ ]:
thresholds = [0.1,0.2,0.25,0.3,0.35,0.4,0.45,0.5,0.6,0.65,0.7,0.75,0.8,0.85,0.9]
thresholds_2 = thresholds[:]
thresholds_2.append('time')

In [ ]:
print(thresholds_2)

In [ ]:
result_model_f1 = pd.DataFrame(index=thresholds_2)
print(result_model_f1)

In [ ]:
start = time.time()
print('start time:',time.strftime('%Y-%m-%d %H:%M:%S',time.localtime(time.time())))

gradient_boosting_classifier = GradientBoostingClassifier()
gradient_boosting_classifier.fit(X_train,y_train.values.ravel())

y_pred = gradient_boosting_classifier.predict(X_test.values)

# Plot non-normalized confusion matrix
plt.figure()
# plot_confusion_matrix(y_test,y_pred,title='Confusion matrix')

end = time.time()
print(end-start,'s')

different threshold

In [ ]:
y_pred_proba = gradient_boosting_classifier.predict_proba(X_test.values)

In [ ]:
result_model_f1['GradientBoostingClassifier'] = 0
# print(result_model_f1)

In [ ]:
for i in thresholds:
    y_test_predictions_high_recall = y_pred_proba[:,1] > i
    result_model_f1.loc[i,'GradientBoostingClassifier'] = f1_score(y_test.values,y_test_predictions_high_recall) #记录f1

In [ ]:
result_model_f1.loc['time','GradientBoostingClassifier'] = end-start
print(result_model_f1)

In [ ]:
from sklearn.externals import joblib

print(gradient_boosting_classifier)
joblib.dump(gradient_boosting_classifier, 'gradient_boosting_classifier.model')

Regression

In [2]:
import pandas as pd
import copy
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection  import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
import time

C:\Users\pcaoaa\AppData\Local\Continuum\anaconda3\lib\site-packages\sklearn\ensemble\weight_boosting.py:29: DeprecationWarning: numpy.core.umath_tests is an internal NumPy module and should not be imported. It will be removed in a future NumPy release.
  from numpy.core.umath_tests import inner1d


In [3]:
data = pd.read_csv("tap_fun_train_7_pay.csv", index_col=0, parse_dates=True)

In [4]:
data_pay_more = copy.copy(data[data['pay_price']<data['prediction_pay_price']])
data_pay_more.shape

(11309, 111)

In [5]:
data_pay_more = data_pay_more.drop(['user_id'],axis=1)
data = copy.copy(data_pay_more)
data.shape

(11309, 110)

In [6]:
Quantity = 'prediction_pay_price'

In [7]:
X = data.loc[:, data.columns != Quantity]
y = data.loc[:, data.columns == Quantity]

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.22, random_state = 0)

In [9]:
print("Number transactions train dataset: ", len(X_train))
print("Number transactions test dataset: ", len(X_test))
print("Total number of transactions: ", len(X_train)+len(X_test))

100% data
Number transactions train dataset:  8821
Number transactions test dataset:  2488
Total number of transactions:  11309


In [10]:
start = time.time()
print('start time:',time.strftime('%Y-%m-%d %H:%M:%S',time.localtime(time.time())))

gradient_boosting_regression = GradientBoostingRegressor()
gradient_boosting_regression.fit(X_train,y_train.values.ravel())

y_pred = gradient_boosting_regression.predict(X_test.values)

# The mean squared error
print("Root Mean squared error: %.2f"
      % mean_squared_error(y_test, y_pred) ** 0.5)
# Explained variance score: 1 is perfect prediction
print('Variance score: %.2f' % r2_score(y_test, y_pred))

end = time.time()
print(end-start,'s')

start time: 2018-10-30 14:23:42
Root Mean squared error: 976.82
Variance score: 0.40
2.4888529777526855 s


In [11]:
data = pd.read_csv("tap_fun_test.csv", parse_dates=True)
data.shape #(828934, 108)

(828934, 108)

In [12]:
object_columns_df = data.select_dtypes(include=["object"])

In [13]:
data['register_time_month'] = data.register_time.str[5:7]
data['register_time_day'] = data.register_time.str[8:10]
data = data.drop(['register_time'],axis=1) 

In [14]:
data[['register_time_month','register_time_day']] = data[['register_time_month','register_time_day']].apply(pd.to_numeric)

In [15]:
data['register_time_count'] = data['register_time_month'] * 31 + data['register_time_day'] 

In [16]:
# save users who paid within first 7 days
data_7_pay = copy.copy(data[data['pay_price']>0])
print(data_7_pay.shape)  # (19549, 110)
data_7_pay.to_csv ("tap_fun_test_7_pay.csv")

(19549, 110)


In [17]:
# save users who did not pay within first 7 days
data_7_NOT_pay = copy.copy(data[data['pay_price']==0])
print(data_7_NOT_pay.shape)  #(809385, 110)
data_7_NOT_pay.to_csv ("tap_fun_test_7_NOT_pay.csv")

(809385, 110)


In [18]:
data_test = pd.read_csv("tap_fun_test_7_NOT_pay.csv", index_col=0, parse_dates=True)
print(data_test.shape)# (809385, 110) 

(809385, 110)


Users who did not pay for the first 7 days would not pay in the future

In [19]:
data_test_part1 = data_test[['user_id','pay_price']]
data_test_part1.rename(columns={'pay_price':'prediction_pay_price'}, inplace = True)
data_test_part1.to_csv('tap_fun_test_part1_still0.csv')

C:\Users\pcaoaa\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\frame.py:3781: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  return super(DataFrame, self).rename(**kwargs)


In [20]:
data_test = pd.read_csv("tap_fun_test_7_pay.csv", index_col=0, parse_dates=True)
print(data_test.shape)  # (19549, 110)

(19549, 110)


In [27]:
data_test_model = data_test.drop([ 'user_id'],axis=1)

In [28]:
from sklearn.externals import joblib
gradient_boosting_classifier = joblib.load('gradient_boosting_classifier.model')

In [30]:
y_test_pred = gradient_boosting_classifier.predict(data_test_model.values)

In [31]:
# transform to DataFrame
y_test_pred = pd.DataFrame(y_test_pred, columns= {'pred_label'})

In [34]:
columns_test = data_test.columns
data_test = data_test.values
data_test = pd.DataFrame(data_test, columns = columns_test )

In [35]:
y_test_pred = pd.concat([data_test, y_test_pred], axis=1)

In [36]:
y_test_pred.shape # (19549, 111)

(19549, 111)

* 1: only paid within 7 days
* 0: continued to pay

In [37]:
y_test_pred['pred_label'].value_counts()

1    15587
0     3962
Name: pred_label, dtype: int64

* predict that users who only paid within first 7 days would pay the same amount within 45 days
* keep users who continued to pay for regression

In [38]:
y_test_pred_part2 = copy.copy(y_test_pred[y_test_pred['pred_label']==1])
y_test_pred_part3 = copy.copy(y_test_pred[y_test_pred['pred_label']==0])

In [39]:
y_test_pred_part2_user_id = pd.DataFrame(y_test_pred_part2,columns ={'user_id'})
y_test_pred_part2_pay = pd.DataFrame(y_test_pred_part2,columns ={'pay_price'})
y_test_pred_part2 = pd.concat([y_test_pred_part2_user_id, y_test_pred_part2_pay], axis=1)
y_test_pred_part2.rename(columns={'pay_price':'prediction_pay_price'}, inplace = True)

In [40]:
y_test_pred_part2.to_csv('tap_fun_test_part2_nopaymore.csv')
y_test_pred_part3.to_csv('tap_fun_test_part3_paymore.csv')

regression: users who continued to pay

In [53]:
y_test_pred_part3 = pd.read_csv("tap_fun_test_part3_paymore.csv", index_col=0, parse_dates=True)
y_test_pred_part3.shape #(3962, 111)

(3962, 111)

In [54]:
user_id_pay_more = y_test_pred_part3['user_id'].values
user_id_pay_more[:10]

array([15403., 15491., 15492., 15841., 15960., 17034., 17542., 17757.,
       18328., 18524.])

In [55]:
y_test_pred_part3 = y_test_pred_part3.drop(['user_id'],axis=1)
y_test_pred_part3 = y_test_pred_part3.drop(['pred_label'],axis=1)

In [57]:
y_test_pred_part3_howmuch = gradient_boosting_regression.predict(y_test_pred_part3.values)

In [58]:
y_test_pred_part3_user_id = pd.DataFrame(user_id_pay_more,columns = {'user_id'})
y_test_pred_part3_howmuch = pd.DataFrame(y_test_pred_part3_howmuch,columns = {'prediction_pay_price'})

In [59]:
y_test_pred_part3 = pd.concat([y_test_pred_part3_user_id, y_test_pred_part3_howmuch], axis=1)
y_test_pred_part3.shape  # (3962, 2)

(3962, 2)

In [60]:
y_test_pred_part3.to_csv('tap_fun_test_part3_paymore_result.csv')

Combine results of the three parts

In [2]:
pred_part1 = pd.read_csv("tap_fun_test_part1_still0.csv", index_col=0, parse_dates=True)
print(pred_part1.shape) #(809385, 2)

(809385, 2)


In [3]:
pred_part2 = pd.read_csv("tap_fun_test_part2_nopaymore.csv", index_col=0, parse_dates=True)
print(pred_part2.shape) #(15587, 2)

(15587, 2)


In [4]:
pred_part3 = pd.read_csv("tap_fun_test_part3_paymore_result.csv", index_col=0, parse_dates=True)
print(pred_part3.shape) #(3962, 2)

(3962, 2)


In [5]:
pred = pred_part1.append(pred_part2)
pred = pred.append(pred_part3)
pred.shape # (828934, 2)

(828934, 2)

In [6]:
pred.to_csv('result.csv', index=False)